presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -577,15 +581,21 @@ public boolean transitionToFinishing()
         }
 
         if (autoCommit) {
-            transactionManager.asyncCommit(session.getTransactionId().get())
-                    .whenComplete((value, throwable) -> {
-                        if (throwable == null) {
-                            transitionToFinished();
-                        }
-                        else {
-                            transitionToFailed(unwrapCompletionException(throwable));
-                        }
-                    });
+            ListenableFuture<?> commitFuture = transactionManager.asyncCommit(session.getTransactionId().get());
+            Futures.addCallback(commitFuture, new FutureCallback<Object>()
+            {
+                @Override
+                public void onSuccess(@Nullable Object result)
+                {
+                    transitionToFinished();
+                }
+
+                @Override
+                public void onFailure(Throwable throwable)
+                {
+                    transitionToFailed(unwrapCompletionException(throwable));
+                }
+            });
```

Comment:
Thanks for taking time to share your opinion. As to the API, time will show what we need, but I feel we will need something :)